### ISSUE TITLE
support for high bit depth images

### ISSUE THREAD

**Issue by yellowtailfan** [openseadragon/openseadragon#1643]
I'd like to extend OSD to support high bit depth images (as discussed in #1482) for scientific image processing. I'm opening this issue to get feedback on what I'm proposing to do, and on how to implement it.

The features I'd like are (most important first):
1. Support uint8, uint16, float16 and float32.
2. Support tone mapping from high bit depth down to 8-bit sRGB for display (range mapping, gamma, pseudo-colour, etc.).
3. Support multi-channel images (not just 1 or 3 channel, but N-channel).
4. Support tone mapping using multiple channels (e.g. for visualising complex floating point images represented as a real and imaginary pair of channels, show the modulus as luminance, and the phase as hue).

So far I can see 2 possible approaches:
1. Use EXR image tiles, via the emscripten WebAssembly OpenEXR wrapper from the [JERI viewer](https://github.com/disneyresearch/jeri), plus GLSL for tone mapping to 8-bit sRGB tiles.
2. Use PNG image tiles, with 2-byte (uint16 or float16)or 4-byte (float32) scalars packed across 2 (RG) or 4 (RGBA) colour channels, with GLSL for both image decoding and  tone mapping tiles.

I think I will try (1) first. EXR has many features that I think will be useful: it handles uint8, uint16, float16 and float32, multiple channels and image metadata. EXR also has lossless and lossy compression options.

If EXR seems too slow, or has other issues, then I may try option (2). Packing the data into PNG is a bit of a kludge, and while PNG compression is lossless, I think PNG compression is designed for images containing vector-style graphics and might not perform very well when used for lossless compression of natural image data. PNG will also make diagnostics more difficult and will require custom kludges for each new data type and to support multiple channels.

Using (1) does introduce a dependency. I think I will structure my implementation as a plugin to OSD. My plugin would depend on:
* the OpenEXR wrapper from JERI, 
* viaWebGL. 
* and possibly other plugins.

However, I expect I will also need to modify core OSD code, so I may propose such changes as pull requests when the time comes.

**Comment by iangilman** [comment]
Sounds awesome! I agree with your analysis. I don't know that much about EXR vs PNG, but less hacky sounds good. And yes, if it adds a dependency it should be a plugin. Whatever changes are necessary in OSD to support that plugin (and others like it) are welcome.

This is possibly of interest to @thejohnhoffer, @jmuhlich, and @dpwrussell.

**Comment by yellowtailfan** [comment]
I've spent some time studying the internals of OSD and viaWebGL, and I've come up with a rough plan of attack: creating a new extension of `TileSource` called `HDRTileSource`, similar to `ImageTileSource`.

I'll go into more details about my proposal later in this comment, but first I will outline my understanding of the current pipeline.

OSD understandably assumes that images can be decoded natively by the browser, and therefore uses `HTMLImageElement` as the JS type for loading and rendering images. Image loading is triggered asynchronously by assigning a URL to `image.src`, where `image` is created using `new Image()`. The loaded image is cached and then later drawn onto the viewport context using `drawImage()`.

viaWebGL intervenes using the `tile-drawing` event to send the image through GLSL. Currently it overwrites the input cached image with the output from GLSL, but there is an [interactive branch of viaWebGL ](https://github.com/thejohnhoffer/viaWebGL/pull/6) where both the input and output images are cached, so that the GLSL can quickly applied with different parameters (e.g. to adjust image gamma without flickering). The branch as-is doesn't quite work, but I have made a one-line change which fixes it and makes gamma adjustment nice and smooth.

In some cases, `image.src` is assigned to an object or data URL (rather than a network resource URL), such as when the image is loaded using AJAX, or when the image is filtered with GLSL.

To summarise the current image loading and rendering pipeline using viaWebGL:
1. Image URL pointing to 8-bit browser native format (PNG, JPEG, WEBP, etc.)
2. Image is loaded by assigning image.src to the URL (where image is created using `new Image()`)
3. Loaded image is cached
4. Rendered image is created using GLSL, overwriting the loaded image in the cache
5. Cached rendered image is drawn to viewport context

My proposal is to create an alternative pipeline:
1. Image URL pointing to HDR **delivery format** image (EXR containing uint16, float16 or float32 to begin with)
2. Load the image using AJAX and create a native JS array (Uint16Array or Float32Array)
3. **Native array** image is cached
4. **Rendered image** (8-bit RGB) is created from the native array image using GLSL (without overwriting the native array image in the cache)
5. Rendered image is drawn to viewport context

I'd like to test this pipeline and see I can get it working with reasonable performance, so I'm planning to implement it by creating a new extension of `TileSource` called `HDRTileSource`, similar to `ImageTileSource`. Initially it will only support single images (no pyramids).

Once I get that working, I can share it as a branch for comments.

After that, I can look at a more sustainable solution. At the moment I can see 2 options:
1. Continue extending `HDRTileSource` to handle pyramids
2. Transpose the HDR image handling to intercept `image.src` assignments to support `native array` images

Option (1) seems more straightforward to implement, but duplicates the pyramid tile handling.

Option (2) requires generalising the code for image loading, but means that all pyramid tile sources get the option to use HDR image file formats.

I will also consider at that point where to put the code -- which parts should go in OSD and which parts belong in a new plugin.

Thanks for reading my plan! I'd appreciate any comments or feedback.

**Comment by iangilman** [comment]
Sounds good to me! And yes, I would say your understanding of the existing pipeline is correct. 

One thing to think about is whether you would want this pipeline to function for all the tile source types (DZI, IIIF, etc.) or if it should be its own tile source.

**Comment by yellowtailfan** [comment]
I’ve got float EXR images displaying in OSD using the HDRTileSource approach. However, my implementation breaks many things and is totally unsuitable for merging at this point. 

I agree it would be better if all tile sources could benefit from HDR image support. 

The trick is that as far as I can tell a lot of the code which directly handles images assumes that an image is either an HTMLImageElement, or a canvas, or a URL pointing to an HTMLImageElement. It looks like I might need to abstract those points in the code so that both ArrayType and HTMLImageElement can be used. That’s as far as I’ve got with thinking about how to do it.

**Comment by iangilman** [comment]
Yes, that sounds like it would be the cleanest approach. It does sound like a bit of work... Are you up for taking a stab at it?

**Comment by yellowtailfan** [comment]
I think I could do it. It might not be the next thing I get onto though. My HDRTileSource approach is close enough to be useful to my other work, but I do aim to merge a more sustainable solution at some point.

**Comment by yellowtailfan** [comment]
On further thought, I think the point where I will get to implementing a sustainable HDR implementation is when I need tiled HDR support. That would be a good opportunity to refactor what I've done so far.

Meanwhile, to briefly explain what I've done so far, I've used 2 properties:
* imageType (either 'LdrImage' or 'HdrImage') and 
* imageArray (an ArrayType when imageType is 'HdrImage') 

to complement the url/src option used for LDR images.

To get the data from HdrTileSource into the rendered canvas required adding imageType and imageArray to the options for:
* ImageJob
* ImageLoader
* Tile
* HdrTileSource

A more sustainable solution might involve encoding the ArrayTile as a data URL, but I suspect there would still need to be some additional alternative code to handle the 2 image types (e.g. via a new ImageData type).

**Comment by iangilman** [comment]
Cool. Seems like a plan. 

I think as long as we are trying to do it cleanly we probably don't need to make data URLs out of the image arrays; I think it would be better if we just understood that there were different types of images. Anyway, we can cross that bridge when we get to it.

I'm glad you've got something that works for your needs!
